reference/src/taimpl/src/internal/rsa.c

```diff
@@ -18,12 +18,480 @@
 
 #include "rsa.h" // NOLINT
 #include "digest_util.h"
+#include "digest_util_mbedtls.h"
 #include "log.h"
 #include "pkcs8.h"
 #include "porting/memory.h"
 #include "stored_key_internal.h"
+#include "pkcs12_mbedtls.h"
 #include <memory.h>
-#include <openssl/pem.h>
+
+/*
+ * ============================================================================
+ * CUSTOM RSA OAEP IMPLEMENTATION WITH DUAL HASH SUPPORT
+ * ============================================================================
+ * 
+ * Background:
+ * mbedTLS 2.16.10 only supports a single hash algorithm for both OAEP and MGF1
+ * via the ctx->hash_id field. However, PKCS#1 v2.1 allows different hash
+ * algorithms for OAEP label hashing and MGF1 mask generation.
+ * 
+ * This custom implementation is copied from mbedTLS 2.16.10 rsa.c and modified
+ * to support separate OAEP hash and MGF1 hash parameters.
+ * 
+ * Source: mbedTLS 2.16.10 library/rsa.c
+ * Functions: mgf_mask(), mbedtls_rsa_rsaes_oaep_decrypt()
+ * Modified: Added separate hash parameters for OAEP and MGF1
+ * 
+ * Rationale:
+ * While this enables full PKCS#1 v2.1 compliance in software, note that
+ * hardware crypto accelerators in OP-TEE 3.18 deployments may still have
+ * the same limitation. This implementation is primarily for testing and
+ * validation purposes.
+ * ============================================================================
+ */
+
+/**
+ * Custom MGF1 mask generation function (copied from mbedTLS 2.16.10)
+ * 
+ * This is a direct copy of the static mgf_mask function from mbedTLS rsa.c.
+ * No modifications needed - it accepts a pre-configured md_ctx.
```

Comment:
Please add a comment before line 58:
```
which is Copyright The Mbed TLS Contributors
Licensed under the Apache License, Version 2.0
```
Also, please append a credit to the end of NOTICE at top level:
```
Code from mbedTLS is:
Copyright The Mbed TLS Contributors
Licensed under the Apache License, Version 2.0
```